File: server/routes/musetalk.ts
import { Router } from 'express';
import path from 'path';
import fs from 'fs/promises';
import { museTalkService } from '../services/musetalkService';
import { BATTLE_CHARACTERS } from '../../shared/characters';

const router = Router();

/**
 * Initialize MuseTalk service
 * POST /api/musetalk/init
 */
router.post('/init', async (req, res) => {
  try {
    const initialized = await museTalkService.initialize();
    
    res.json({
      success: initialized,
      status: museTalkService.getStatus()
    });
  } catch (error) {
    console.error('MuseTalk initialization error:', error);
    res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Get MuseTalk service status
 * GET /api/musetalk/status
 */
router.get('/status', (req, res) => {
  res.json({
    success: true,
    status: museTalkService.getStatus()
  });
});

/**
 * Prepare character avatar for lip sync
 * POST /api/musetalk/prepare/:characterId
 */
router.post('/prepare/:characterId', async (req, res) => {
  try {
    const { characterId } = req.params;
    
    const character = BATTLE_CHARACTERS.find((c) => c.id === characterId);
    if (!character) {
      return res.status(404).json({
        success: false,
        error: 'Character not found'
      });
    }

    const result = await museTalkService.prepareAvatar(character);
    
    if (result.success) {
      res.json(result);
    } else {
      res.status(400).json(result);
    }
  } catch (error) {
    console.error('Avatar preparation error:', error);
    res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Generate lip-synced video for character
 * POST /api/musetalk/generate
 * Body: { characterId, audioPath, outputPath? }
 */
router.post('/generate', async (req, res) => {
  try {
    const { characterId, audioPath, outputPath } = req.body;

    if (!characterId || !audioPath) {
      return res.status(400).json({
        success: false,
        error: 'characterId and audioPath are required'
      });
    }

    const character = BATTLE_CHARACTERS.find((c) => c.id === characterId);
    if (!character) {
      return res.status(404).json({
        success: false,
        error: 'Character not found'
      });
    }

    // Generate output path if not provided
    const finalOutputPath = outputPath || path.join(
      process.cwd(),
      'server/public/videos',
      `${characterId}_${Date.now()}.mp4`
    );

    // Ensure output directory exists
    await fs.mkdir(path.dirname(finalOutputPath), { recursive: true });

    let result;
    
    if (museTalkService.isAvailable()) {
      // Use MuseTalk for advanced lip sync
      result = await museTalkService.generateLipSyncVideo(
        character,
        audioPath,
        finalOutputPath
      );
    } else {
      // Fallback to static video
      result = await museTalkService.createFallbackVideo(
        character,
        audioPath,
        finalOutputPath
      );
    }

    if (result.success) {
      res.json({
        ...result,
        videoUrl: `/videos/${path.basename(finalOutputPath)}`
      });
    } else {
      res.status(400).json(result);
    }
  } catch (error) {
    console.error('Video generation error:', error);
    res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

/**
 * Generate lip-synced video for battle round
 * POST /api/musetalk/battle/:battleId/:roundId
 */
router.post('/battle/:battleId/:roundId', async (req, res) => {
  try {
    const { battleId, roundId } = req.params;
    const { characterId } = req.body;

    // This would integrate with the battle system to automatically
    // generate lip-synced videos for AI responses
    
    // For now, return a placeholder response
    res.json({
      success: true,
      message: 'Battle video generation not yet implemented',
      battleId,
      roundId,
      characterId
    });
  } catch (error) {
    console.error('Battle video generation error:', error);
    res.status(500).json({
      success: false,
      error: error instanceof Error ? error.message : 'Unknown error'
    });
  }
});

export default router;

File: server/services/musetalk.py
#!/usr/bin/env python3
"""
MuseTalk Integration Service for Real-Time Lip Sync
Provides video lip-sync generation for the rap battle application
"""

import os
import sys
import json
import argparse
import tempfile
import subprocess
import shlex
from pathlib import Path
from typing import Optional, Dict, Any
import base64
import logging

# Add MuseTalk to Python path
MUSETALK_PATH = os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(__file__))), 'MuseTalk')
sys.path.append(MUSETALK_PATH)

try:
    import torch
    import cv2
    import numpy as np
    from omegaconf import OmegaConf
    from transformers import WhisperModel
    
    # Import MuseTalk modules
    from musetalk.utils.face_parsing import FaceParsing
    from musetalk.utils.utils import datagen, load_all_model
    from musetalk.utils.preprocessing import get_landmark_and_bbox, read_imgs
    from musetalk.utils.blending import get_image_prepare_material, get_image_blending
    from musetalk.utils.audio_processor import AudioProcessor
    
    MUSETALK_AVAILABLE = True
except ImportError as e:
    print(f"MuseTalk dependencies not available: {e}")
    MUSETALK_AVAILABLE = False

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class MuseTalkLipSync:
    """Real-time lip sync processor using MuseTalk"""
    
    def __init__(self):
        self.initialized = False
        self.models = {}
        self.avatars = {}
        
        if not MUSETALK_AVAILABLE:
            logger.warning("MuseTalk not available, running in simulation mode")
            return
            
        self.models_dir = os.path.join(MUSETALK_PATH, 'models')
        self.results_dir = os.path.join(MUSETALK_PATH, 'results', 'realtime')
        
        # Ensure directories exist
        os.makedirs(self.results_dir, exist_ok=True)
        
    def initialize_models(self) -> bool:
        """Initialize MuseTalk models for processing"""
        if not MUSETALK_AVAILABLE:
            return False
            
        try:
            logger.info("Initializing MuseTalk models...")
            
            # Check for required model files
            required_models = [
                'musetalkV15/unet.pth',
                'sd-vae/diffusion_pytorch_model.bin', 
                'whisper/pytorch_model.bin',
                'face-parse-bisent/79999_iter.pth'
            ]
            
            missing_models = []
            for model_path in required_models:
                full_path = os.path.join(self.models_dir, model_path)
                if not os.path.exists(full_path):
                    missing_models.append(model_path)
                    
            if missing_models:
                logger.error(f"Missing MuseTalk models: {missing_models}")
                logger.info("Please run: cd MuseTalk && sh download_weights.sh")
                return False
                
            # Load models (this would be the actual MuseTalk model loading)
            # For now, we'll simulate the initialization
            self.models = {
                'unet': f"{self.models_dir}/musetalkV15/unet.pth",
                'vae': f"{self.models_dir}/sd-vae/diffusion_pytorch_model.bin",
                'whisper': f"{self.models_dir}/whisper/pytorch_model.bin",
                'face_parser': f"{self.models_dir}/face-parse-bisent/79999_iter.pth"
            }
            
            logger.info("MuseTalk models initialized successfully")
            self.initialized = True
            return True
            
        except Exception as e:
            logger.error(f"Failed to initialize MuseTalk models: {e}")
            return False
    
    def prepare_avatar(self, image_path: str, avatar_id: str) -> Dict[str, Any]:
        """Prepare avatar for lip sync processing"""
        try:
            if not self.initialized and not self.initialize_models():
                return {"success": False, "error": "MuseTalk not initialized"}
                
            logger.info(f"Preparing avatar {avatar_id} from {image_path}")
            
            # Create avatar directory
            avatar_dir = os.path.join(self.results_dir, 'avatars', avatar_id)
            os.makedirs(avatar_dir, exist_ok=True)
            
            # Copy image to avatar directory
            import shutil
            avatar_image = os.path.join(avatar_dir, 'avatar.png')
            shutil.copy2(image_path, avatar_image)
            
            # For now, simulate avatar preparation
            # In full implementation, this would process the face landmarks, 
            # extract facial features, and prepare for real-time inference
            
            avatar_info = {
                "avatar_id": avatar_id,
                "image_path": avatar_image,
                "prepared": True,
                "bbox_shift": 0,  # This would be calculated from face detection
                "landmarks": []   # This would contain facial landmarks
            }
            
            # Save avatar info
            info_path = os.path.join(avatar_dir, 'avatar_info.json')
            with open(info_path, 'w') as f:
                json.dump(avatar_info, f, indent=2)
                
            self.avatars[avatar_id] = avatar_info
            
            return {
                "success": True, 
                "avatar_id": avatar_id,
                "info": avatar_info
            }
            
        except Exception as e:
            logger.error(f"Avatar preparation failed: {e}")
            return {"success": False, "error": str(e)}
    
    def generate_lipsync_video(self, avatar_id: str, audio_path: str, output_path: str) -> Dict[str, Any]:
        """Generate lip-synced video for avatar with audio"""
        try:
            if avatar_id not in self.avatars:
                return {"success": False, "error": f"Avatar {avatar_id} not prepared"}
                
            logger.info(f"Generating lip sync video for {avatar_id}")
            
            avatar_info = self.avatars[avatar_id]
            avatar_dir = os.path.join(self.results_dir, 'avatars', avatar_id)
            
            # For demonstration, create a simple video with the avatar image
            # In full implementation, this would use MuseTalk's real-time inference
            
            # Get audio duration
            result = subprocess.run([
                'ffprobe', '-v', 'quiet', '-show_entries', 
                'format=duration', '-of', 'csv=p=0', audio_path
            ], capture_output=True, text=True)
            
            try:
                duration = float(result.stdout.strip())
            except:
                duration = 5.0  # Default duration
                
            # Create video from static image with audio
            # This is a placeholder - real MuseTalk would generate animated frames
            cmd = [
                'ffmpeg', '-y',
                '-loop', '1', '-i', avatar_info['image_path'],
                '-i', audio_path,
                '-c:v', 'libx264', '-tune', 'stillimage',
                '-c:a', 'aac', '-b:a', '192k',
                '-pix_fmt', 'yuv420p',
                '-shortest',
                '-t', str(duration),
                output_path
            ]
            
            result = subprocess.run(cmd, capture_output=True, text=True)
            
            if result.returncode != 0:
                logger.error(f"FFmpeg error: {result.stderr}")
                return {"success": False, "error": "Video generation failed"}
                
            # In real implementation, this would be the actual MuseTalk processing:
            # 1. Extract audio features using Whisper
            # 2. Generate lip sync frames using UNet model
            # 3. Blend frames with original avatar
            # 4. Encode final video
                
            return {
                "success": True,
                "video_path": output_path,
                "duration": duration,
                "method": "placeholder_static"  # Would be "musetalk_realtime" in full implementation
            }
            
        except Exception as e:
            logger.error(f"Video generation failed: {e}")
            return {"success": False, "error": str(e)}

def main():
    """CLI interface for MuseTalk service"""
    parser = argparse.ArgumentParser(description='MuseTalk Lip Sync Service')
    parser.add_argument('command', choices=['init', 'prepare', 'generate'])
    parser.add_argument('--avatar-id', help='Avatar identifier')
    parser.add_argument('--image-path', help='Path to avatar image')
    parser.add_argument('--audio-path', help='Path to audio file')
    parser.add_argument('--output-path', help='Output video path')
    
    args = parser.parse_args()
    
    service = MuseTalkLipSync()
    
    if args.command == 'init':
        result = service.initialize_models()
        print(json.dumps({"success": result}))
        
    elif args.command == 'prepare':
        if not args.avatar_id or not args.image_path:
            print(json.dumps({"success": False, "error": "avatar-id and image-path required"}))
            sys.exit(1)
            
        result = service.prepare_avatar(args.image_path, args.avatar_id)
        print(json.dumps(result))
        
    elif args.command == 'generate':
        if not all([args.avatar_id, args.audio_path, args.output_path]):
            print(json.dumps({"success": False, "error": "avatar-id, audio-path, and output-path required"}))
            sys.exit(1)
            
        result = service.generate_lipsync_video(args.avatar_id, args.audio_path, args.output_path)
        print(json.dumps(result))

if __name__ == '__main__':
    main()